browser-specific-failures.js

```diff
@@ -0,0 +1,386 @@
+'use strict';
+
+/**
+ * Implements a view of how many browser specific failures each engine has over
+ * time.
+ */
+
+const fetch = require ('node-fetch');
+const fs = require ('fs');
+const flags = require('flags');
+const Git = require('nodegit');
+const lib = require('./lib');
+const moment = require('moment');
+
+flags.defineString('from', '2018-07-01', 'Starting date (inclusive)');
+flags.defineString('to', moment().format('YYYY-MM-DD'), 'Ending date (exclusive)');
+flags.defineStringList('products', ['chrome', 'firefox', 'safari'], 'Browsers to compare. Must match the products used on wpt.fyi');
+flags.defineString('output', null, 'Output CSV file to write to. Defaults to {stable, experimental}-browser-specific-failures.csv');
+flags.defineBoolean('experimental', false, 'Calculate metrics for experimental runs.');
+flags.parse();
+
+// Helper function that counts how many test failures there are within a given
+// set of status results.
+function countFailures(statuses) {
+  let count = 0;
+  for (const testStatus of statuses) {
+    if (testStatus == 'FAIL') {
+      count += 1;
+    }
+  }
+  return count;
+}
+
+// Produce a 'score' of browser-specific failures for a given set of runs from
+// different products on the same WPT codebase. The word 'score' is used instead
+// of count as we normalize the counts of subtests.
+//
+// Returns a map from product name to score.
+function scoreBrowserSpecificFailures(runs, expectedBrowsers) {
```

Comment:
I haven't reviewed this in detail, I want to give some ideas for the "may take >5 minutes" problem first.

The `run.tree` in `runs` come from `lib.results.getGitTree(repo, run)` and any duplicate trees or subtrees will be same very same object. It ought to be very common that parts of the subtrees are identical from one day to the next.

To take advantage of this to speed this up, this (or a helper) should probably be structured as a taking n trees and work by recursively walking n subtrees. Whether you walk the union of the trees or the intersection of the trees depends on how you want to treat missing results.

Then, you should be able to see a massive speedup with simple memoization: if called again with the same n trees, just returned a cached value.

WDYT? This might be fast enough as it is, so you decide if it's worth the investment.